### PR TITLE
[Bugfix][Intel] Fix XPU Dockerfile Build

### PR DESCRIPTION
**Pull request by tylertitsworth**
The XPU Dockerfile consistently wasn't building, I've worked out some of the issues and assigned the correct IPEX XPU versions to be installed in the image.

I also added support for the openai server version of xpu. I validated on a Max 1100.

Because Ubuntu 20.04 isn't supported by the Triton XPU Backend anymore, 22.04 needs to be utilized for the newer python version.

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

File: Dockerfile.xpu
Comment (tylertitsworth) [inline Dockerfile.xpu:1]: If we could do without full oneapi it would be nice to utilize `intel/intel-optimized-pytorch:2.1.40-xpu-pip-base` instead and reduce the image size by 12GB.

**Comment by liuxingbin** [comment]
Hi
sorry to bother you. I use the ubuntu-22.04 as the base docker-images, but I met the problem 
`torch.xpu.availble=False, torch.xpu.device_count()=0`
I follow the install doc strictly: https://docs.vllm.ai/en/latest/getting_started/xpu-installation.html.
BTW,  I use 'clinfo' to check the hardware device,  which shows intel-gpu is on my computer.

Any advice for my problem?

**Comment by yma11** [comment]
@tylertitsworth I used the changed Dockerfile as proposed in this PR and image can be built successfully. But I met following error when starting the server:
```
ImportError: /usr/local/lib/python3.10/dist-packages/intel_extension_for_pytorch/lib/libxetla_kernels.so: undefined symbol: _ZN4sycl3_V17handler22setKernelIsCooperativeEb

```
Does it happen in your env?

File: Dockerfile.xpu
Comment (yma11) [inline Dockerfile.xpu:26]: Are these common requirements all required for xpu?

**Comment by tylertitsworth** [comment]
> Does it happen in your env?

I didn't encounter this issue. I'm using a Flex GPU fyi.

File: Dockerfile.xpu
Comment (tylertitsworth) [reply Dockerfile.xpu:26]: They are required for vLLM AFAIK.

File: requirements-common.txt
Comment (tylertitsworth) [reply requirements-common.txt:8]: On a prior version tiktoken threw an error for needing a version under 20. Now it works so I've uncapped the version.

**Comment by yma11** [comment]
> > Does it happen in your env?
> 
> I didn't encounter this issue. I'm using a Flex GPU fyi.

I am using `Intel(R) Arc(TM) A770 Graphics` and the error always can be reproduced.

**Comment by tylertitsworth** [comment]
> I am using Intel(R) Arc(TM) A770 Graphics and the error always can be reproduced.

Let me test on that GPU and see if I can figure out a fix.

**Comment by tylertitsworth** [comment]
> ImportError: /usr/local/lib/python3.10/dist-packages/intel_extension_for_pytorch/lib/libxetla_kernels.so: undefined symbol: _ZN4sycl3_V17handler22setKernelIsCooperativeEb

I was able to reproduce, even on the Flex system and found that it was a oneapi issue that prevented vLLM from importing ipex.

I bumped the oneapi base image version to 2024.2.0 and it resolved the issue.

**Comment by tylertitsworth** [comment]
@WoosukKwon Can I get a maintainer to help review and approve?

**Comment by tylertitsworth** [comment]
@jikunshang

**Comment by youkaichao** [comment]
I will delegate the review to @jikunshang 

feel free to ping for review/merge after @jikunshang 's approval

**Comment by jikunshang** [comment]
hi @tylertitsworth thanks for your contribution. I have verified on PVC, seems it doesn't work. when I try to run `example/offline_inference.py`, it throw an error like `AttributeError: module 'intel_extension_for_pytorch.llm.functional' has no attribute 'varlen_attention'`. As far as I know, this api is not shown on ipex-xpu 2.1.4 release [note](https://github.com/intel/intel-extension-for-pytorch/releases/tag/v2.1.40%2Bxpu). But this will be supported in ipex-xpu 2.3 release. 

Can you double check whether this could run vllm engine on your environment?

File: requirements-xpu.txt
Comment (jikunshang) [inline requirements-xpu.txt:7]: `torchvision` and `torchaudio` are unnecessary

**Comment by tylertitsworth** [comment]
> Can you double check whether this could run vllm engine on your environment?

@jikunshang I get the same error. This will likely be solved once ipex xpu is bumped to 2.3.0. I added the server implemented in lieu of the support being added. If you just build the vllm-base target do you receive the same error?

**Comment by jikunshang** [comment]
> > Can you double check whether this could run vllm engine on your environment?
> 
> @jikunshang I get the same error. This will likely be solved once ipex xpu is bumped to 2.3.0. I added the server implemented in lieu of the support being added. If you just build the vllm-base target do you receive the same error?

I think docker build (for current code base) issue is due to docker image version, [here ](https://github.com/vllm-project/vllm/pull/6517/files#diff-edcc1d6789abce643f02737a7a8b235ff0a396b806372251a8e831242e1ee405R1)change to ubuntu 20.04, revert to 22.04 would resolve.

For ipex 2.1.40, I think this is not target for vllm support, thanks for your contribution of fixing these build issue.

For ipex 2.3, I am collabrating with ipex team to integrate and validate. I will update vllm to latest once ipex 2.3 get released.

**Comment by tylertitsworth** [comment]
@jikunshang I updated based on your feedback, please let me know if something else needs to be attended to.

File: requirements-xpu.txt
Comment (jikunshang) [inline requirements-xpu.txt:4]: I feel we should keep this.

File: Dockerfile.xpu
Comment (jikunshang) [inline Dockerfile.xpu:1]: Please use 2024.2.1

**Review by jikunshang — APPROVED** [review]

**Review by jikunshang — APPROVED** [review]

**Comment by jikunshang** [comment]
LGTM. @youkaichao Please help merge, thanks a lot!

File: requirements-xpu.txt
Comment (tylertitsworth) [reply requirements-xpu.txt:4]: https://github.com/advisories/GHSA-cx63-2mw6-8hw5

**Comment by tylertitsworth** [comment]
> LGTM. @youkaichao Please help merge, thanks a lot!

@liuxingbin thank you for your review, @youkaichao have you had a chance to check on this PR?

**Review by jikunshang — APPROVED** [review]

File: vllm/platforms/__init__.py
Comment (youkaichao) [inline vllm/platforms/__init__.py:49]: does `torch.xpu.is_available()` initialize some xpu context? I find `torch.cuda.is_available()` will initialize cuda context and will be harmful for tensor parallel.

**Review by youkaichao — APPROVED** [review]
please address the comments

File: vllm/platforms/__init__.py
Comment (tylertitsworth) [reply vllm/platforms/__init__.py:49]: Based on the [source](https://github.com/pytorch/pytorch/blob/df1eef9779bb07fbe86aba63fa413a62885c0400/torch/csrc/api/src/xpu.cpp#L6-L12) for this function I believe it does not initialize any context, and simply checks the device count.

**Comment by tylertitsworth** [comment]
> please address the comments

Hi @youkaichao, I have addressed the comment.

File: Dockerfile.xpu
Comment (youkaichao) [inline Dockerfile.xpu:26]: the intel gpu test fails because of this.

File: requirements-xpu.txt
Comment (jikunshang) [inline requirements-xpu.txt:17]: I feel we should remove this. it will throw error if we execute `VLLM_TARGET_DEVICE=xpu python setup.py install`

File: Dockerfile.xpu
Comment (tylertitsworth) [reply Dockerfile.xpu:26]: @youkaichao I have restored it.

**Review by jikunshang — APPROVED** [review]

File: Dockerfile.xpu
Comment (dtrifiro) [inline Dockerfile.xpu:32]: `--no-cache-dir` defeats the purpose of having `--mount=type=cache` above.

**Pull request merged**
